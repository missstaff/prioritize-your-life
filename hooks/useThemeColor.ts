import { useColorScheme } from "react-native";
import { COLORTHEME } from "@/constants/Colors";


export function useThemeColor(
  props: { light?: string; dark?: string },
  colorName: keyof typeof COLORTHEME.light & keyof typeof COLORTHEME.dark
) {
  const theme = useColorScheme() ?? "light";
  const colorFromProps = props[theme];

  if (colorFromProps) {
    return colorFromProps;
  } else {
    return COLORTHEME[theme][colorName];
  }
}
