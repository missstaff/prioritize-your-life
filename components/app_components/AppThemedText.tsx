import { Text, type TextProps } from "react-native";
import { ScaledSheet, s } from "react-native-size-matters";
import { useThemeColor } from "@/hooks/useThemeColor";
import { COLORS, COLORTHEME } from "@/constants/Colors";

export type ThemedTextProps = TextProps & {
  type?: "default" | "title" | "defaultSemiBold" | "subtitle" | "link";
};

/**
 * A custom text component that applies theme-based styling.
 * @param style - The style object to be applied to the text component.
 * @param lightColor - The color to be used when the theme is light.
 * @param darkColor - The color to be used when the theme is dark.
 * @param type - The type of text styling to be applied.
 * @param rest - Additional props to be passed to the text component.
 * @returns {JSX.Element} The rendered AppThemedText component.
 */
export function AppThemedText({
  style,
  type = "default",
  ...rest
}: ThemedTextProps): JSX.Element {
  const color = useThemeColor(
    { light: COLORTHEME.light.text, dark: COLORTHEME.dark.text },
    "text"
  );

  return (
    <Text
      style={[
        { color },
        type === "default" ? styles.default : undefined,
        type === "title" ? styles.title : undefined,
        type === "defaultSemiBold" ? styles.defaultSemiBold : undefined,
        type === "subtitle" ? styles.subtitle : undefined,
        type === "link" ? styles.link : undefined,
        style,
      ]}
      {...rest}
    />
  );
}

const styles = ScaledSheet.create({
  default: {
    fontSize: s(16),
    lineHeight: s(24),
  },
  defaultSemiBold: {
    fontSize: s(16),
    fontWeight: "600",
    lineHeight: s(24),
  },
  title: {
    fontSize: s(32),
    fontWeight: "bold",
    lineHeight: s(32),
  },
  subtitle: {
    fontSize: s(20),
    fontWeight: "bold",
  },
  link: {
    color: COLORS.primary,
    fontSize: s(16),
    lineHeight: s(30),
  },
});
