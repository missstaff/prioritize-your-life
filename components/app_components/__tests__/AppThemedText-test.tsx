import React from "react";
import renderer from "react-test-renderer";
import { AppThemedText } from "../AppThemedText";
import { useThemeColor } from "@/hooks/useThemeColor";

describe("AppThemedText Tests", () => {
  
  afterEach(() => {
    jest.clearAllMocks();
  });

  it("renders correctly with default type and light theme", () => {
    (useThemeColor as jest.Mock).mockReturnValue("#000000");

    const tree = renderer
      .create(<AppThemedText>Default Text</AppThemedText>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders correctly with title type and dark theme", () => {
    (useThemeColor as jest.Mock).mockReturnValue("#ffffff");

    const tree = renderer
      .create(<AppThemedText type="title">Title Text</AppThemedText>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders correctly with defaultSemiBold type", () => {
    (useThemeColor as jest.Mock).mockReturnValue("#000000");

    const tree = renderer
      .create(
        <AppThemedText type="defaultSemiBold">SemiBold Text</AppThemedText>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders correctly with subtitle type", () => {
    (useThemeColor as jest.Mock).mockReturnValue("#000000");

    const tree = renderer
      .create(<AppThemedText type="subtitle">Subtitle Text</AppThemedText>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders correctly with link type", () => {
    (useThemeColor as jest.Mock).mockReturnValue("#000000");

    const tree = renderer
      .create(<AppThemedText type="link">Link Text</AppThemedText>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders correctly with custom styles", () => {
    (useThemeColor as jest.Mock).mockReturnValue("#000000");

    const tree = renderer
      .create(
        <AppThemedText style={{ fontSize: 24, fontWeight: "bold" }}>
          Custom Styled Text
        </AppThemedText>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
