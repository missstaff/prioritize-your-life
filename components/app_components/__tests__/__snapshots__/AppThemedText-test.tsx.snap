// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppThemedText Tests renders correctly with custom styles 1`] = `
<Text
  style={
    [
      {
        "color": "#000000",
      },
      {
        "fontSize": 34.285714285714285,
        "lineHeight": 51.42857142857143,
      },
      undefined,
      undefined,
      undefined,
      undefined,
      {
        "fontSize": 24,
        "fontWeight": "bold",
      },
    ]
  }
>
  Custom Styled Text
</Text>
`;

exports[`AppThemedText Tests renders correctly with default type and light theme 1`] = `
<Text
  style={
    [
      {
        "color": "#000000",
      },
      {
        "fontSize": 34.285714285714285,
        "lineHeight": 51.42857142857143,
      },
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
    ]
  }
>
  Default Text
</Text>
`;

exports[`AppThemedText Tests renders correctly with defaultSemiBold type 1`] = `
<Text
  style={
    [
      {
        "color": "#000000",
      },
      undefined,
      undefined,
      {
        "fontSize": 34.285714285714285,
        "fontWeight": "600",
        "lineHeight": 51.42857142857143,
      },
      undefined,
      undefined,
      undefined,
    ]
  }
>
  SemiBold Text
</Text>
`;

exports[`AppThemedText Tests renders correctly with link type 1`] = `
<Text
  style={
    [
      {
        "color": "#000000",
      },
      undefined,
      undefined,
      undefined,
      undefined,
      {
        "color": "#CD03EF",
        "fontSize": 34.285714285714285,
        "lineHeight": 64.28571428571428,
      },
      undefined,
    ]
  }
>
  Link Text
</Text>
`;

exports[`AppThemedText Tests renders correctly with subtitle type 1`] = `
<Text
  style={
    [
      {
        "color": "#000000",
      },
      undefined,
      undefined,
      undefined,
      {
        "fontSize": 42.857142857142854,
        "fontWeight": "bold",
      },
      undefined,
      undefined,
    ]
  }
>
  Subtitle Text
</Text>
`;

exports[`AppThemedText Tests renders correctly with title type and dark theme 1`] = `
<Text
  style={
    [
      {
        "color": "#ffffff",
      },
      undefined,
      {
        "fontSize": 68.57142857142857,
        "fontWeight": "bold",
        "lineHeight": 68.57142857142857,
      },
      undefined,
      undefined,
      undefined,
      undefined,
    ]
  }
>
  Title Text
</Text>
`;
